go/vt/vtctl/workflow/sequences.go

```diff
@@ -0,0 +1,697 @@
+	for _, sm := range sequences {
+		initGroup.Go(func() error {
+			currentVal, err := ts.getCurrentSequenceValue(gctx, sm)
+			if err != nil {
+				return err
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			currentValues[sm.backingTableName] = currentVal
+			return nil
+		})
+	}
+	errs := initGroup.Wait()
+	if errs != nil {
+		return nil, errs
+	}
+	return currentValues, nil
+}
+
+func (ts *trafficSwitcher) getCurrentSequenceValue(ctx context.Context, seq *sequenceMetadata) (int64, error) {
+	if err := seq.escapeValues(); err != nil {
+		return 0, err
+	}
+	sequenceShard, ierr := ts.TopoServer().GetOnlyShard(ctx, seq.backingTableKeyspace)
+	if ierr != nil || sequenceShard == nil || sequenceShard.PrimaryAlias == nil {
+		return 0, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the primary tablet for keyspace %s: %v",
+			seq.backingTableKeyspace, ierr)
+	}
+	sequenceTablet, ierr := ts.TopoServer().GetTablet(ctx, sequenceShard.PrimaryAlias)
+	if ierr != nil || sequenceTablet == nil {
+		return 0, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the primary tablet for keyspace %s: %v",
+			seq.backingTableKeyspace, ierr)
+	}
+	if sequenceTablet.DbNameOverride != "" {
+		seq.backingTableDBName = sequenceTablet.DbNameOverride
+	}
+	query := sqlparser.BuildParsedQuery(sqlGetCurrentSequenceVal,
+		seq.backingDB,
+		seq.backingTable,
+	)
+	qr, ierr := ts.ws.tmc.ExecuteFetchAsApp(ctx, sequenceTablet.Tablet, true, &tabletmanagerdatapb.ExecuteFetchAsAppRequest{
+		Query:   []byte(query.Query),
+		MaxRows: 1,
+	})
+	if ierr != nil || len(qr.Rows) != 1 {
+		return 0, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the current value for the sequence table %s.%s: %v",
+			seq.backingTableDBName, seq.backingTableName, ierr)
+	}
+	rawVal := sqltypes.Proto3ToResult(qr).Rows[0][0]
+	if rawVal.IsNull() {
+		return 0, nil
+	}
+	currentVal, err := rawVal.ToInt64()
+	if err != nil {
+		return 0, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to convert the current value for the sequence table %s.%s: %v",
+			seq.backingTableDBName, seq.backingTableName, err)
+	}
+	return currentVal, nil
+}
+
+func (ts *trafficSwitcher) updateSequenceValue(ctx context.Context, seq *sequenceMetadata, currentMaxValue int64) error {
+	if err := seq.escapeValues(); err != nil {
+		return err
+	}
+	nextVal := currentMaxValue + 1
+	log.Infof("Updating sequence %s.%s to %d", seq.backingTableDBName, seq.backingTableName, nextVal)
+
+	// Now we need to update the sequence table, if needed, in order to
+	// ensure that the next value it provides is > the current max.
+	sequenceShard, ierr := ts.TopoServer().GetOnlyShard(ctx, seq.backingTableKeyspace)
+	if ierr != nil || sequenceShard == nil || sequenceShard.PrimaryAlias == nil {
+		return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the primary tablet for keyspace %s: %v",
+			seq.backingTableKeyspace, ierr)
+	}
+	sequenceTablet, ierr := ts.TopoServer().GetTablet(ctx, sequenceShard.PrimaryAlias)
```

Comment:
Not necessary to do here, but noting that we could save the tablets used in `getCurrentSequenceValue` and re-use them here so that we don't double the topo calls. If the total time taken is a concern, this refactor will make that problem worse (though topo calls should be fast vs mysqld calls).

Reply:
Agree, I didn't want to do that here because this was already a big refactor and didn't want to add more changes to this PR and try to keep the flow the same as the current one.

There are various performance improvements possible which I am planning to do soon, : https://github.com/vitessio/vitess/issues/17871